src/main/java/edu/ucla/library/prl/harvester/Institution.java

```diff
@@ -0,0 +1,375 @@
+        myContactMethods = Objects.requireNonNull(aContactMethods);
+        myWebsite = Objects.requireNonNull(aWebsite);
+    }
+
+    /**
+     * Instantiates an institution from its JSON representation.
+     * <p>
+     * <b>This constructor is meant to be used only by generated service proxy code!</b>
+     * {@link #Institution(String, String, String, ContactMethods, URL)} should be used everywhere else.
+     *
+     * @param aJsonObject An institution represented as JSON
+     * @throws InvalidInstitutionJsonException If the JSON representation is invalid
+     */
+    @SuppressWarnings({ "PMD.CognitiveComplexity", "PMD.CyclomaticComplexity", "PMD.NPathComplexity" })
+    public Institution(final JsonObject aJsonObject) {
+        Objects.requireNonNull(aJsonObject);
+
+        final String name = aJsonObject.getString(NAME);
+        final String description = aJsonObject.getString(DESCRIPTION);
+        final String location = aJsonObject.getString(LOCATION);
+        final Optional<InternetAddress> email;
+        final Optional<PhoneNumber> phone;
+        final Optional<URL> webContact;
+        final String website = aJsonObject.getString(WEBSITE);
+
+        if (name != null) {
+            myName = name;
+        } else {
+            throw new InvalidInstitutionJsonException(MessageCodes.PRL_002, NAME);
+        }
+
+        if (description != null) {
+            myDescription = description;
+        } else {
+            throw new InvalidInstitutionJsonException(MessageCodes.PRL_002, DESCRIPTION);
+        }
+
+        if (location != null) {
+            myLocation = location;
+        } else {
+            throw new InvalidInstitutionJsonException(MessageCodes.PRL_002, LOCATION);
+        }
+
+        email = Optional.ofNullable(aJsonObject.getString(EMAIL)).map(rawEmail -> {
+            try {
+                return new InternetAddress(rawEmail, true);
+            } catch (final AddressException details) {
+                throw new InvalidInstitutionJsonException(details, MessageCodes.PRL_004, EMAIL, details.getMessage());
+            }
+        });
+        phone = Optional.ofNullable(aJsonObject.getString(PHONE)).map(rawPhone -> {
+            try {
+                return PHONE_NUMBER_UTIL.parse(rawPhone, null);
+            } catch (NumberParseException details) {
+                throw new InvalidInstitutionJsonException(details, MessageCodes.PRL_004, PHONE, details.getMessage());
+            }
+        });
+        webContact = Optional.ofNullable(aJsonObject.getString(WEB_CONTACT)).map(rawWebContact -> {
+            try {
+                return new URL(rawWebContact);
+            } catch (MalformedURLException details) {
+                throw new InvalidInstitutionJsonException(details, MessageCodes.PRL_004, WEB_CONTACT,
+                        details.getMessage());
+            }
+        });
+
+        if (email.isPresent() && phone.isPresent() && webContact.isPresent()) {
```

Comment:
Sorry to step in this late... but do you really need all the specialized `ContactMethods` constructors? Could you do a `if (isPresent()) Optional.of() else Optional.empty()` for each param in one constructor?

Reply:
Here's the thread where we discussed this: https://github.com/UCLALibrary/prl-harvester/pull/10#discussion_r955041182

Long story short is that using ContactMethods provides a more type-safe API (i.e., only calls to `Institution(String, String, String, ContactMethods, URL)` constructor that don't throw an exception [will compile](https://github.com/UCLALibrary/prl-harvester/pull/10#discussion_r955301809)).

Reply:
Ahh, looking at this again and now I think it's better to have the three Optional params (with the body checking that at least one is not empty). I didn't realize earlier how the many constructors for ContactMethods is going to require the use of a conditional like [this](https://github.com/UCLALibrary/prl-harvester/pull/10/files#diff-0b11154b96c3ba34289ef2a1ad9eba504c55d444bdca2d2b05b1c72031f11bd8R172-R188) to figure out which one to call (whenever the state of the cron timers must be restored from the database on app restart, for example).